query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandler.java

```diff
@@ -74,13 +88,70 @@ public QueryCost canHandle(QueryRequest request, ExecutionContext executionConte
   @Override
   public Observable<Row> handleRequest(
       QueryRequest originalRequest, ExecutionContext executionContext) {
-
     // Validate QueryContext and tenant id presence
     Preconditions.checkNotNull(executionContext);
     Preconditions.checkNotNull(executionContext.getTenantId());
 
-    // todo call convert and execute request using client here
+    Map<Request, PromQLMetricResponse> responseMap;
+    Map<String, String> logicalAttributeNameToMetricQueryMap = new LinkedHashMap<>();
+    if (isRangeQueryRequest(originalRequest)) {
+      PromQLRangeQueries promQLRangeQueries =
+          requestToPromqlConverter.convertToPromqlRangeQuery(
+              executionContext,
+              originalRequest,
+              executionContext.getAllSelections(),
+              logicalAttributeNameToMetricQueryMap);
+      responseMap = prometheusRestClient.executeRangeQuery(promQLRangeQueries);
+    } else {
+      PromQLInstantQueries promQLInstantQueries =
+          requestToPromqlConverter.convertToPromqlInstantQuery(
+              executionContext,
+              originalRequest,
+              executionContext.getAllSelections(),
+              logicalAttributeNameToMetricQueryMap);
+      responseMap = prometheusRestClient.executeInstantQuery(promQLInstantQueries);
+    }
+
+    List<Row> rows =
+        PrometheusBasedResponseBuilder.buildResponse(
+            responseMap,
+            prometheusViewDefinition.getAttributeMap(),
+            logicalAttributeNameToMetricQueryMap,
+            prepareSelectionColumnSet(executionContext.getAllSelections(), executionContext),
+            executionContext.getTimeFilterColumn());
+
+    return Observable.fromIterable(rows).doOnNext(row -> LOG.debug("collect a row: {}", row));
+  }
+
+  private boolean isRangeQueryRequest(QueryRequest queryRequest) {
+    return queryRequest.getGroupByList().stream().anyMatch(QueryRequestUtil::isDateTimeFunction);
+  }
 
-    return null;
+  private LinkedHashSet<String> prepareSelectionColumnSet(
+      LinkedHashSet<Expression> expressions, ExecutionContext executionContext) {
+    LinkedHashSet<String> selectionColumnSet = new LinkedHashSet<>();
```

Comment:
collect and make immutable rather than returning an immutable type. Also, why are we returning `LinkedHashSet` If we're concerned about order and need to declare it, return a list; if we're not concerned about order or we are in this function but not elsewhere, return a `Set`. In other words, does this api need to _both_ declare that it's unique and ordered, or are those implementation details?

Reply:
The requirement from this method is that the returned collection has unique element and the order is same as the input collection. 

Now returning set from this method would mean losing the order. 
A list can be returned as it will maintain the order (the input collection already has unique elements so will the list)

If you see the flow of data,
this method is invoked on a collection in the ExecutionContext class, which returns object of type LinkedHashSet, 
then that collection is processed by this method. Finally it is passed into the response builder class a method param. 
```
PrometheusBasedResponseBuilder.buildResponse(
            responseMap,
            prometheusViewDefinition.getAttributeMap(),
            logicalAttributeNameToMetricQueryMap,
            prepareSelectionColumnSet(executionContext.getAllSelections(), executionContext),
            executionContext.getTimeFilterColumn());
```

Checking the users of `executionContext.getAllSelections()`, I think this collection itself could have been of `List` type

I understand that wherever possible we should return interface, but how to convey that the param `LinkedHashSet<String> columnSelectionSet` (which will be changed to List<String> columnSelectionSet) in this method org.hypertrace.core.query.service.prometheus.PrometheusBasedResponseBuilder#buildResponse must have ordered and unique elements?

Reply:
> I understand that wherever possible we should return interface, but how to convey that the param LinkedHashSet<String> columnSelectionSet (which will be changed to List<String> columnSelectionSet) in this method org.hypertrace.core.query.service.prometheus.PrometheusBasedResponseBuilder#buildResponse must have ordered and unique elements?

The problem here isn't just "use an interface instead of an implementation", it's what that rule of thumb is based on - that returning the type is leaking implementation details. It's much less about the implementation of the collection itself, it's that we're trying to convey the responsibilities of one class to another with it. We only need (or want) one class to understand how to prepare selections, and that's the only place that needs to know if duplicates are allowed or not. The caller can then trust that prepare the selection column set was responsible for doing everything involved in preparing a selection set, but it doesn't need to know what that was.